File: src/routes/analytics.ts

import { Router } from 'express';
import { getAggregateInsights } from '../services/aggregationService';
import logger from '../utils/logger';
import Papa from 'papaparse';

const router = Router();

// GET /api/analytics/:dataset?year=2025&format=csv
router.get('/:dataset', async (req, res) => {
    try {
        const { dataset } = req.params;
        const { year, format } = req.query;
        
        const validDatasets = ['biometric', 'enrolment', 'demographic'];
        if (!validDatasets.includes(dataset)) {
            return res.status(400).json({ error: 'Invalid dataset' });
        }

        const data = await getAggregateInsights(dataset, year as string);
        
        if (format === 'csv') {
            // Flatten logic for Looker Studio
            // We'll create a "Category, Label, Count" schema for maximum flexibility
            // OR simply "State, Age Group, Count" (but they are separate aggregates in our service).
            
            // Better: Return "State Wise" first as it's the main map visual
            // Looker typically needs one table. Since we have TWO insights (State vs Age),
            // maybe we default to STATE here, or allow ?view=age
            
            const view = req.query.view || 'state';
            let rows: any[] = [];
            
            if (view === 'state') {
                 rows = Object.entries(data.by_state).map(([state, count]) => ({
                     State: state,
                     Updates: count
                 }));
            } else if (view === 'age') {
                 rows = Object.entries(data.by_age_group).map(([age, count]) => ({
                     AgeGroup: age,
                     Updates: count
                 }));
            }

            const csv = Papa.unparse(rows);
            res.header('Content-Type', 'text/csv');
            res.attachment(`${dataset}_${view}_${year || 'all'}.csv`);
            return res.send(csv);
        }

        res.json({
            dataset,
            year: year || 'all',
            generated_at: new Date().toISOString(),
            data
        });

    } catch (err: any) {
        logger.error('Analytics Error', err);
        res.status(500).json({ error: err.message });
    }
});

export default router;

File: app/api/v1/endpoints/analytics.py
from fastapi import APIRouter, Response, HTTPException
from fastapi.responses import JSONResponse
from app.services.aggregation_service import get_aggregate_insights
from app.utils.logger import get_logger
from datetime import datetime
import csv
import io

logger = get_logger()
router = APIRouter()

@router.get("/{dataset}")
async def get_analytics(dataset: str, year: str = None, format: str = None, view: str = 'state'):
    try:
        valid_datasets = ['biometric', 'enrolment', 'demographic']
        if dataset not in valid_datasets:
            return Response(content='{"error": "Invalid dataset"}', status_code=400, media_type="application/json")
            
        data = await get_aggregate_insights(dataset, year)
        
        if format == 'csv':
            # Flatten Logic from Node
            rows = []
            if view == 'state':
                 for state, count in data['by_state'].items():
                     rows.append({"State": state, "Updates": count})
            elif view == 'age':
                 for age, count in data['by_age_group'].items():
                     rows.append({"AgeGroup": age, "Updates": count})
                     
            output = io.StringIO()
            writer = csv.DictWriter(output, fieldnames=rows[0].keys() if rows else [])
            writer.writeheader()
            writer.writerows(rows)
            
            headers = {
                'Content-Disposition': f'attachment; filename="{dataset}_{view}_{year or "all"}.csv"'
            }
            return Response(content=output.getvalue(), media_type="text/csv", headers=headers)
            
        headers = {
            "Cache-Control": "public, s-maxage=3600, stale-while-revalidate=600",
            "X-Generated-At": datetime.now().isoformat()
        }
        
        return JSONResponse(
            content={
                "dataset": dataset,
                "year": year or 'all',
                "generated_at": datetime.now().isoformat(),
                "data": data
            },
            headers=headers
        )
        
    except ValueError as e:
        # e.g. Dataset file not found
        return Response(content=f'{{"error": "{str(e)}"}}', status_code=500, media_type="application/json")
    except Exception as e:
        logger.error(f"Analytics Error: {e}")
        return Response(content=f'{{"error": "{str(e)}"}}', status_code=500, media_type="application/json")
